add code comment style guide for TiKV (#4204)
tikv/tikv: code-comment-style.md

# Code Comment Style

This document describes the code comment style applied to TiKV repositories. When you are to commit, be sure to follow the style to write good code comments.

## Why a good comment matters?

- To speed up the reviewing process
- To help maintain the code
- To improve the API document readability
- To improve the development efficiency of the whole team

## Where/When to comment?

Write a comment where/when:

- For important code
- For obscure code
- For tricky or interesting code
- For a complex code block
- If a bug exists in the code but you cannot fix it or you just want to ignore it for the moment
- If the code is not optimal but you don’t have a smarter way now
- To remind yourself or others of missing functionality or upcoming requirements not present in the code

A comment is generally required for:

> - File

Reviewer: Prefer module to file, since files are a secondary concept in Rust

- Type
- Constant
- Function
- Method
- Variable
- Typical algorithm
- Exported name

> - Test case

Reviewer: My preference is not to require comments anywhere since that leads to boilerplate. Use comments when there is context that is missing from the code or is hard to deduce from the code. For example, what are the side-effects of calling a function, or the performance characteristics, or the use cases which the code addresses.

As an explicit counter-example I think comments like:

```
/// Returns a record from the database.
pub fn get_record(db: &Database) -> Record { ... }
```

are a negative and should be avoided, better to have no comment at all.

A good comment is something like:

```
/// This function may get the record from disk which means it will block waiting for IO. It is nearly always
/// better to call `async_get_record` unless you are sure that the requested record is cached in memory.
///
/// If you need to get many records, calling this function many times has O(n) performance. Consider using
/// `get_records` which has O(log n) performance.
///
/// If `db` is in an inconsistent state, this function will panic.
pub fn get_record(db: &Database) -> Record { ... }
```

Note that we don't say what `get_record` does but describe aspects of how it should be used which are not obvious from the function signature.

Author: Good point. Will add in style guide. We should not comment for comment purposes, but comment to deliver something that matters

- TODO
- FIXME

## How to comment?

### Format of a good comment (for Rust)

- Non-doc comment

    - Used for inline code comment (for better code explaination when reviewing)
    - Use **//** for a line comment

>     - Use **/* … */** for a block comment (used only when needed)

Reviewer: I would never use block comments

Author: Not contradicted with what's implied by "used only when needed"

    
- Doc comment

    - Used in module level descriptions (structures, fields, macros, etc.) for command line reference
    - Use **///** for a line comment 
    - Use **//!** for a block comment

>     - Use **rustfmt** to format your code

Reviewer: I would not use rustfmt to format comments - it is not stable for comments and formats imperfectly

Author: Noted.

Author: @zhangjinpeng1987 PTAL

Author: Here **rustfmt** is used to format code, not the comment, so will leave as is for now.

- Place the single-line and block comment above the code it’s annotating
- Fold long lines of comments

> - Each line of text in your code and comment should be at most 100 characters long

Reviewer: I prefer 80, 100 is fine for code, but I think is a long line of text to read (studies show fewer words per column are easier to read).

Reviewer 2: I suggest to keep the same value as [style guide](https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/guide.md#indentation-and-line-width). And I prefer 100, which works better for my large monitor. :rofl:

Author: Updated as suggested by Jay.

- Do not use relative URLs; use **absolute URLs** instead

    Right:
    
    ```
    // For more configuration details, see https://example.com/project/tools/config-items.
    ```
    
    Wrong:
    
    ```
    // For more details, see [Configuration Items](../tools/config-items.md).
    ```

### Language for a good comment

- Word
    
    - Use **American English** rather than British English
        
        - color, canceling, synchronize     (Recommended)
        - colour, cancelling, synchronise   (Not recommended)
    
    - Use correct spelling

    - Use **standard or official capitalization**
        
        - TiKV, TiDB-Binlog, Region, gRPC, RocksDB, GC, k8s, [mydumper](https://github.com/maxbube/mydumper), [Prometheus Pushgateway](https://github.com/prometheus/pushgateway)   (Right)
        - Tikv, TiDB Binlog, region, grpc, rocksdb, gc, k8S, MyDumper, Prometheus PushGateway   (Wrong)

    - Use words and expressions consistently
        
        - dead link vs. broken link (Only one of them can appear in a single document)
    
    - Do not use lengthy compound words

>     - Do not abbreviate unless it is absolutely necessary

Reviewer: I think abreviations can be fine if they improve readability, but avoid overusing them. For example, it would be tedious to read 'Titanium Data Base' all the time instead of 'TiDB', or to spell out 'HTTP'

Author: True. Since this is code comment we do not have to follow this rule as strictly as documentation. For general accepted/known abbreviations, we can and should use to improve readability. For less-known items, we may still need to spell out

>     - "we" can be used only when it means the code writer

Reviewer: Should the author always prefer 'we' to 'I'? (I think yes)

Author: Agree. It would be self-pretentious to use "I".

- Sentence

    - Use standard grammar and correct punctuation
    - Use relatively short sentences

- For each comment, capitalize the first letter and end this sentence with a period

- When used for description, comments should be **descriptive** rather than **imperative**

    - Opens the file   (Right)
    - Open the file    (Wrong)       

- Use "this" instead of "the" to refer to the current thing
    
    - Gets the toolkit for this component   (Recommended)
    - Gets the toolkit for the component    (Not recommended)

- The Markdown format is allowed
    
    - Opens the `log` file  

### Tips for a good comment

- Comment code while writing it
- Do not assume the code is self-evident
- Avoid unnecessary comments for simple code
- Write comments as if they were for you

> - Make sure the comment is up-to-date

Reviewer: I would also add "Make sure you keep comments up to date when you edit code"

> # Code Comment Style

Reviewer 3: Maybe this file need to be capitalized like other md files in the project root?

Author: Good point. Will update accordingly.

This document describes the code comment style applied to TiKV repositories. When you are to commit, be sure to follow the style to write good code comments.

> ## Why a good comment matters?

Reviewer 4: ```suggestion
## Why does a good comment matter?
```

- To speed up the reviewing process
- To help maintain the code
- To improve the API document readability
- To improve the development efficiency of the whole team

## Where/When to comment?

Write a comment where/when there is context that is missing from the code or is hard to deduce from the code. Specifically, use a comment:

- For important code
- For obscure code
- For tricky or interesting code
- For a complex code block
- If a bug exists in the code but you cannot fix it or you just want to ignore it for the moment
- If the code is not optimal but you don’t have a smarter way now
- To remind yourself or others of missing functionality or upcoming requirements not present in the code

A comment is generally required for:

- Module
- Type
- Constant
- Function
- Method
- Variable
- Typical algorithm
- Exported name
- Test case
- TODO
- FIXME

## How to comment?

### Format of a good comment (for Rust)

- Non-doc comment

    - Used to document implementation details
    - Use **//** for a line comment
    - Use **/* … */** for a block comment (used only when needed)
    
- Doc comment

    - Used to document the interface of code (structures, fields, macros, etc.)

>     - Use **///** for a line comment 

Reviewer 4: ```suggestion
    - Use **///** for item documentation (functions, attributes, structures, etc)
```

>     - Use **//!** for a block comment

Reviewer 4: ```suggestion
    - Use **//!** for module level documentation```

Reviewer 4: (There is also block comments but according to @brson they are bad style. They are `/*! */`.)

>     - Use **rustfmt** to format your code

Reviewer 4: Not sure this is appropriate for a comment style document. Rustfmt is only for code.

Author: Decided to remove this after discussion with Catin, since it does not pertain to comment style guide.

- Place the single-line and block comment above the code it’s annotating
- Fold long lines of comments
- The maximum width for a line is 100 characters

> - Do not use relative URLs; use **absolute URLs** instead

Reviewer 4: It is absolutely appropriate to use relative URLs in autogenerated documentation. The user's machine will generate them via `cargo doc` and the hosting location may be anything. So using an absolute URL will either break local documentation, or hosted documentation.

For example, this is how you link to another struct: https://github.com/tikv/client-rust/blob/master/src/raw.rs#L75

Author: As confirmed with @BusyJay, we should use relative URLs for local links.  Now updated as: Use "Use relative URLs when you link to positions on your local machine"

    Right:
    
    ```
    // For more configuration details, see https://example.com/project/tools/config-items.
    ```
    
    Wrong:
    
    ```
    // For more details, see [Configuration Items](../tools/config-items.md).
    ```

### Language for a good comment

- Word
    
    - Use **American English** rather than British English
        
        - color, canceling, synchronize     (Recommended)
        - colour, cancelling, synchronise   (Not recommended)
    
    - Use correct spelling

    - Use **standard or official capitalization**
        
        - TiKV, TiDB-Binlog, Region, gRPC, RocksDB, GC, k8s, [mydumper](https://github.com/maxbube/mydumper), [Prometheus Pushgateway](https://github.com/prometheus/pushgateway)   (Right)
        - Tikv, TiDB Binlog, region, grpc, rocksdb, gc, k8S, MyDumper, Prometheus PushGateway   (Wrong)

    - Use words and expressions consistently
        
        - dead link vs. broken link (Only one of them can appear in a single document)
    
    - Do not use lengthy compound words

    - Do not abbreviate unless it is necessary (for readability purposes)

>     - "we" can be used only when it means the code writer

Reviewer 4: ```suggestion
    - "we" should be used only when it means the code writer *and* the reader.
```

# Code Comment Style

This document describes the code comment style applied to TiKV repositories. When you are to commit, be sure to follow the style to write good code comments.

## Why a good comment matters?

- To speed up the reviewing process
- To help maintain the code
- To improve the API document readability
- To improve the development efficiency of the whole team

## Where/When to comment?

Write a comment where/when:

- For important code
- For obscure code
- For tricky or interesting code
- For a complex code block
- If a bug exists in the code but you cannot fix it or you just want to ignore it for the moment
- If the code is not optimal but you don’t have a smarter way now
- To remind yourself or others of missing functionality or upcoming requirements not present in the code

A comment is generally required for:

- File
- Type
- Constant 
- Function
- Method
- Variable
- Typical algorithm
- Exported name
- Test case
- TODO
- FIXME

## How to comment?

### Format of a good comment (for Rust)

> - Non-doc comment

Reviewer 5: When to use non-doc comment and when not to use?

Author: As a general thumb of rule, use doc comment as long as you want to refer to that information in the command line. In our case, we have the convention to use doc comment for (public or private) structures including the fields, macros, methods, etc.  We use non-doc (code) comment for everything else like inner steps of functions.

Will add this in the comment style guide.

Reviewer: I would state the distinction as "use doc comments to document the interface of code and use non-doc comments to document the implementation details"

        
    - Use **//** for a line comment
    - Use **/* … */** for a block comment (used only when needed)
    
- Doc comment
        
    - Use **///** for a line comment
    - Use **//!** for a block comment
    - Use **rustfmt** to format your code

- Place the single-line and block comment above the code it’s annotating
- Fold long lines of comments
- Each line of text in your code and comment should be at most 100 characters long
- Do not use relative URLs; use **absolute URLs** instead

    Right:
    
    ```
    // For more configuration details, see https://example.com/project/tools/config-items.  
    ```
    
    Wrong:
    
    ```
    // For more details, see [Configuration Items](../tools/config-items.md).
    ```

### Language for a good comment

- Word
    
    - Use **American English** rather than British English
        
        - color, canceling, synchronize     (Recommended)
        - colour, cancelling, synchronise   (Not recommended)
    
    - Use correct spelling

    - Use **standard or official capitalization**
        

>         - TiKV, TiDB-Binlog, Region, gRPC, RocksDB, GC, k8s, [mydumper](https://github.com/maxbube/mydumper), [Prometheus Pushgateway](https://github.com/prometheus/pushgateway)   (Right)

Reviewer 6: K8s
not k8s

Reviewer 6: https://kubernetes.io/docs/concepts/overview/what-is-kubernetes/#what-does-kubernetes-mean-k8s

Author: Good spot. Let's stick with K8s in our case.

Reviewer 7: @siddontang @dcalvin “Kubernetes” was stylized as “K8s” formerly; but now it is “k8s”.

Author: As later confirmed with CC, k8s in all lower-cased form is generally accepted. Only initial-capped as K8s when it is the first word of the sentence.

Author: As later confirmed with CC, k8s in all lower-cased form is generally accepted. Only initial-capped as K8s when it is the first word of the sentence.

        - Tikv, TiDB Binlog, region, grpc, rocksdb, gc, K8S, MyDumper, Prometheus PushGateway   (Wrong)

    - Use words and expressions consistently
        
        - dead link vs. broken link (Only one of them can appear in a single document)
    
    - Do not use lengthy compound words

    - Do not abbreviate unless it is absolutely necessary

    - "we" can be used only when it means the code writer

- Sentence

    - Use standard grammar and correct punctuation
    - Use relatively short sentences

- For each comment, capitalize the first letter and end this sentence with a period
    
    - If a lower-case identifier comes at the beginning of a sentence, don't capitalize it

    ```

>     // enterGame causes Players to enter the 

Reviewer 6: For Rust, it is not recommended to use the function name as the start word in comments.

Author: Got it. I can add a note to it.  Here it does not specifically talk about function name. An identifier is a generic word that refers to any construction element like a variable, function, or method. In rust, can we use other names as the start words besides function?

Author: Will remove this since as confirmed by CC, this (identifier) is actually more intended for  Go.

Reviewer: I would add that I would always use markdown code for names of functions, etc., so that if you *have* to start a sentence with a name, then it looks more reasonable, e.g.,

```
// `enter_game` causes Player to enter the ...
fn enter_game ...
```
